### PR TITLE
fix: OTEL context lost in subthreads of parallel bulk calls

### PR DESCRIPTION
**Pull request by claudinoac**
When using OpenTelemetry to trace ES requests, all the parallel bulk chunks go out of the scope of the main span, and that happens because they lose the parent context as they become new threads (which happens in parallel bulk chunks), resulting in what we see below:

Screenshot 2024-07-15 at 18 53 27

In order to keep the same context flowing through all subthreads, I'm using the context propagation concept (https://opentelemetry.io/docs/languages/python/propagation/) to reestablish the previous context on the new spawned thread for every subthread of parallel bulk, thus keeping these subthreads following the flow of the parent context along with all the other calls (search/delete/index/etc).

Here's an example for the same batch job after applying this fix:

Screenshot 2024-07-15 at 18 40 11

I tried to keep the scope to the OTEL class itself and limit the test to the level of the call to recover the context, though I'm aware that might not be the ideal scenario.

I look forward for reviews and suggestions :)

### CODE REVIEW

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
Thank you! Excellent find. I can take a proper look next week, but in the meantime:

1. lint is [still failing](https://github.com/elastic/elasticsearch-py/actions/runs/10099002741/job/27954203164?pr=2616). You can reproduce locally with `nox -rs lint`.
2. The integration tests are also failing with `AttributeError: 'FailingBulkClient' object has no attribute '_otel'`. You can't see those CI logs right now, sorry. It's possible but annoying to reproduce locally with `nox -rs test-3.12` and `nox -rs test_otel-3.12`. But you need an Elasticsearch cluster listening at `localhost:9200`. (I can fix this next week myself, no worries.)

**Comment by claudinoac** [comment]
@pquentin I fixed the lint issues, it seemed wrong bc I was running the linter on python v3.8. 

Now I'm not being able to run the integration tests properly as it's throwing a bunch of SSL errors such as `ValueError: TLS options require scheme to be 'https'`.

Another error that is happening in quite a lot on the tests is `BadRequestError(400, 'illegal_argument_exception', 'Cannot delete policy [kibana-event-log-policy]. It is in use by one or more indices: [.kibana-event-log-8.3.2-000012, .kibana-event-log-8.3.2-000011, .kibana-event-log-8.3.2-000010, .kibana-event-log-8.3.2-000013]')`. 

The ES cluster is empty, configured with basic dev recommended configs and running in a docker container as well as the lib enviroment (I've put them under the same network and changed all the `localhost` references to `{{container_name}} just to run the tests`.  

259 tests are succeeding but other 177 are failing with the errors above

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
buildkite test this please

**Review by pquentin — CHANGES_REQUESTED** [review]
Sorry for the delay, I was on vacation last week. OK, so now that the CI is fixed, I took the time to look at the code in more detail. I'm not familiar with OpenTelemetry contexts and propagation: I just read the documentation today! So please tell me if what I'm saying does not make sense.

My main concern is that the propagation will only behave correctly if there's a context to recover from, but that is only true if one span was created already. If the first function you're calling is `parallel_bulk`, there won't be a span or context. Thankfully, this can be fixed by having the `parallel_bulk` call itself and all callers of `_process_bulk_chunk` create a span, which is something we want anyway. (Bonus points for covering `elasticsearch/_async/helpers.py` too.)

In other words, all bulk helpers should start with something like this:

```python3
with client._otel.span("parallel_bulk", inject_context=True):
    ...
```

File: elasticsearch/_otel.py [hunk 1]
```diff
@@ -86,3 +94,16 @@ def span(
...
+        token = None
+        if self.current_context:
+            otel_parent_ctx = TraceContextTextMapPropagator().extract(
+                carrier=self.current_context
+            )
+            token = otel_context.attach(otel_parent_ctx)
+        yield
+
+        if token:
+            otel_context.detach(token)
```
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:109]: but if there's no token, this call will fail. We still need to check if it exists.

File: elasticsearch/_otel.py [hunk 2]
```diff
@@ -86,3 +94,16 @@ def span(
                 endpoint_id=endpoint_id,
                 body_strategy=self.body_strategy,
             )
+
+    @contextlib.contextmanager
+    def recover_parent_context(self) -> Generator[None, None, None]:
+        token = None
+        if self.current_context:
```
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:101]: I've just put it there as a guard as I thought this could be reused somewhere else.

File: elasticsearch/_otel.py [hunk 3]
```diff
@@ -41,6 +45,8 @@
 
 
 class OpenTelemetry:
+    current_context: dict[str, str] = {}
```
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:48]: I've only put this here to declare the variable, and that default value is needed bc the linter is asking for it.

File: elasticsearch/_otel.py [hunk 4]
```diff
@@ -74,6 +80,8 @@ def span(
 
         span_name = endpoint_id or method
         with self.tracer.start_as_current_span(span_name) as otel_span:
+            self.current_context = {}
```
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:83]: Sure we can remove this as this classes seems to be a singleton

File: elasticsearch/_otel.py [hunk 5]
```diff
@@ -74,6 +80,8 @@ def span(
...
+            TraceContextTextMapPropagator().inject(self.current_context)
```
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:84]: This is literally inside the first span call context.

File: elasticsearch/_otel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:101]: Actually, if there's no active span, and this condition is not present, the call will definitely fail as the carrier will be empty.

File: elasticsearch/_otel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:48]: But now that I'm not initializing it on every span, and considering this is a singleton, it makes more sense to have it declared here this way.

File: elasticsearch/_otel.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (claudinoac) [reply elasticsearch/_otel.py:84]: So what do you suggest? A new parameter to indicate whether we need to dump the context or not to the carrier?

File: elasticsearch/_otel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pquentin) [reply elasticsearch/_otel.py:48]: Right, having `self.current_context: dict[str, str] = {}` in the `__init__` method should be enough to make mypy happy.

File: elasticsearch/_otel.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pquentin) [reply elasticsearch/_otel.py:84]: Yes, I mentioned `inject_context` in my top-level review: https://github.com/elastic/elasticsearch-py/pull/2616#pullrequestreview-2210196802

File: elasticsearch/_otel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pquentin) [reply elasticsearch/_otel.py:101]: My top-level review explains how you can do to always have an active span here: https://github.com/elastic/elasticsearch-py/pull/2616#pullrequestreview-2210196802

**Comment by pquentin** [comment]
Hey @claudinoac, did you get the chance to look at my latest comments?

**Comment by claudinoac** [comment]
@pquentin I did take a look at your observations, and I've been working on them. Though I'm not sure how to proceed with the optional save of context inside the carrier as the span call is made inside the `perform_request` call, which has a few layers behing the bulk call itself. Alternatively, I could detect that a bulk call is being made from the `method` attribute on the span itself, but that would rather be a very coupled and specific solution instead of a more generic one. 
What would you suggest for that instead? Send a flag all the way from the bulk call to the span call, changing all the contracts in between the layers?

**Comment by claudinoac** [comment]
For now, I'll be submitting the changes we agreed to make so far, and in the meanwhile we can discuss the optional saving of context on the carrier.

**Comment by claudinoac** [comment]
Ok @pquentin so I've made the latest changes based on your suggestions, though I wasn't sure what to put into `endpoint_id` and `path_parts` attributes, so I've just put default values there as there wasn't a simple way to infer the endpoint being called (as this is a very generic method) and I'm not exactly sure what `path_parts` mean.

**Comment by pquentin** [comment]
Thanks. I will be able to take a proper look at this next week

**Comment by pquentin** [comment]
Sorry for the delay. This looks pretty good, but now we need to add a span to all helper functions that end up calling `_process_bulk_chunk`, not only `parallel_bulk`. That way, all bulk function will have helpful spans, even if the top-level application does not 't have them. And we need a better way to add that span, using `client._otel.span` is a bit of hack that does not really fit. A quick POC on what this would look like for parallel bulk:

image

I'll work on getting this finished if that's OK for you.

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
The current version is inline with my vision. I now need to add more robust tests and possibly do the same thing for async helpers.

**Comment by pquentin** [comment]
buildkite test this please

**Review by pquentin — APPROVED** [review]

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
buildkite test this please

**Comment by pquentin** [comment]
buildkite test this please

File: elasticsearch/helpers/actions.py [hunk 6]
```diff
@@ -322,6 +324,7 @@ def _process_bulk_chunk(
             Tuple[_TYPE_BULK_ACTION_HEADER, _TYPE_BULK_ACTION_BODY],
         ]
     ],
+    otel_span: OpenTelemetrySpan,
```
Comment (miguelgrinberg) [inline elasticsearch/helpers/actions.py:327]: On a first look it seemed to me this would be optional, to handle the case of apps not using otel. But I guess when otel is disabled this is just a fake span object that doesn't do anything?

**Review by miguelgrinberg — APPROVED** [review]
LGTM

File: elasticsearch/helpers/actions.py (hunk 6 above)
Comment (pquentin) [reply elasticsearch/helpers/actions.py:327]: Yes, exactly, see https://github.com/elastic/elasticsearch-py/pull/2616/files#diff-78ec49b32fc64dc78b569e87b525b22c8d412e606e476c4861d8b108a9289e04R93. Having a fake object avoids using ifs, allowing None, etc.

**Comment by pquentin** [comment]
@claudinoac Sorry for the delay! Yesterday, Miguel made me realize that we don't need context propagation at all here since spans are serializables across threads. But using context propagation was also harmful: if two threads were calling `parallel_bulk` at the same time, then resulting span hierarchy would have been incorrect. I've implemented the simpler and more correct version today. Thanks for your patience.

**Comment by claudinoac** [comment]
@pquentin Yeah I'm aware about them being serializable, in fact that's the way I handle them across threads, processes and even across different applications. Though I'm not sure if your approach here is going to fix the bug. I'm going to make a few live tests just to make sure it addresses the right problem. And for the record, if the context propagation is indeed a problem in race condition scenarios, then the OTEL class shouldn't be a singleton at all.

**Comment by pquentin** [comment]
I can wait for your tests before releasing 8.15.1 if you prefer.

**Comment by claudinoac** [comment]
@pquentin I've made the live tests and everything looks good. Thanks for getting this merged, you can proceed with the release :)
